### DISCUSSION TITLE
Should I switch from MBED core?

### DISCUSSION THREAD

**Discussion by Laserjones** [earlephilhower/arduino-pico#543] [Q&A]
Hi all! Hi Earle and thanks for your great work.

I've recently started programming the Pico in PlatformIO, and so far I've been using the MBED core, because it's officially supported and easy to install. But I keep stumbling across Earle's core when researching for Pico-specific stuff like LittleFS, timer interrupts, or hardware PWM. I've been able to find third-party libraries for MBED for these things, but my impression is that Earle's core is better integrated and - most of all - better documented in these respects. And if there's something I don't like, it's bad or absent documentation.  ;-)

So my honest question to all who are reading this is: Should I switch to this core in spite of incomplete PlatformIO support and the lack of official Arduino.cc support? My existing code is still short and I could surely rewrite it.

My main concerns are these two:

- Will this core continue to evolve and receive support in spite of the official competition?
- Will my code be compatible with other RP2040/MBED-based boards as long as no board-specific features are used?

Any thoughts on this are appreciated, particularly from people who have worked/dealt with both cores.

Kind regards,
Joerg

**Comment by earlephilhower** [comment]
I'll leave others to debate the merits of either implementation, but I do want to clarify one thing.
>Will this core continue to evolve and receive support in spite of the official competition?

I wouldn't call it a competition so much as choice.  I'm just trying to get people coding and making neat things, which is more or less Arduino's mission statement.  Use [this core](https://github.com/earlephilhower/arduino-pico), [MBED](https://github.com/arduino/ArduinoCore-mbed), [CircuitPython](https://circuitpython.org/), [uLisp](http://www.ulisp.com/show?3KN3), or the [raw SDK](https://github.com/raspberrypi/pico-sdk) depending on your skill and goals.

**Comment by Laserjones** [comment]
Is it correct that the MBED OS is something like an additional layer between the code and the hardware when using the MBED core? That overhead would be a strong point in favor of Earle's core (as long as I don't need code compatibility across all MBED-compatible boards).

**Reply by earlephilhower** [reply]
Unless you're using something MBED or Arduino-Pico specific, then code compatibility should be pretty easy.  The Arduino API is still the same between the two.

This core defaults to the Pico-SDK default of 125MHz, while the MBED one runs at 133MHz.  You can manually select the clocks thru the menus here to match, but most testing has been done at 125MHz for this core and 133MHz for the MBED one.

For a trivial example, this shows a 10x difference in performance.  I can't explain it other than the MBED having a lot of management to do between successive calls to `loop` or a bug in MBED code.  So don't expect to see that kind of bump!
https://forum.arduino.cc/t/performance-drifting-down-over-time/981515

For a more real-world test using a complete LISP interpreter, about 60% faster performance was seen:
https://github.com/earlephilhower/arduino-pico/issues/453#issuecomment-1028769532

Differences would be the code here is very thin and MBED does have several layers of indirection per operation, CPU clock speed (but here it's slower than MBED!), and a newer toolchain (GCC10 here vs GCC 8 IIRC).  There could be command-line options for GCC differences (I use `-Os` by default, but you  can change that, too, through the menus).

**Reply by Laserjones** [reply]
> Unless you're using something MBED or Arduino-Pico specific, then code compatibility should be pretty easy.

I do use things like hardware timer interrupts and hardware PWM, so it's not just Arduino code.

But performance is important (I'll try some real-time audio stuff), so I'm now strongly inclined to switch to your core, not to mention the documentation aspect. Thanks for the clarifications.

Is there a known reason to use 125 instead of 133 MHz (other than saving some energy)? I didn't find a reason in the Pico SDK documentation.

**Reply by earlephilhower** [reply]
> I do use things like hardware timer interrupts and hardware PWM, so it's not just Arduino code.

`analogWrite` uses the HW PWM in this core (I'd assume the MBED one, too).   So there'd be no difference there I'd imagine.

For timer interrupts, does MBED have a special API?  Using the Pico SDK's `alarm` timer calls is supported here but there's no particular library for them.

As to the clocks, it's just 125 because that's what the boot ROM sets.  Next major revision (the one w/FreeRTOS support if all goes well), will probably bump it to 133.  For now you can set it yourself (or over/underclock).

**Reply by Laserjones** [reply]
So far I used external libraries from Khoi Hoang for timer interrupts and PWM, as my impression was that the MBED core does not directly support it. But I may be wrong. The documentation is sparse and I'm an absolute beginner.

**Reply by khoih-prog** [reply]
Thanks for using my libraries.

The [**RP2040_PWM**](https://github.com/khoih-prog/RP2040_PWM) supports both this core and MBED core. While you have to use different TimerInterrupt libraries for them.

For this core, use this [**RPI_PICO_TimerInterrupt**](https://github.com/khoih-prog/RPI_PICO_TimerInterrupt)
For MBED, use this [**MBED_RPI_PICO_TimerInterrupt**](https://github.com/khoih-prog/MBED_RPI_PICO_TimerInterrupt)

**Comment by Laserjones** [comment]
Any news on official PlatformIO support, by the way?
